src/MessagePack.UnityClient/Assets/Scripts/MessagePack/IFormatterResolver.cs

```diff
@@ -64,6 +66,8 @@ private static void Throw(Type t, IFormatterResolver resolver)
             throw new FormatterNotRegisteredException(t.FullName + " is not registered in resolver: " + resolver.GetType());
         }
 
+        private static readonly ConcurrentDictionary<IFormatterResolver, ThreadsafeTypeKeyHashTable<object>> FormatterCache = new ConcurrentDictionary<IFormatterResolver, ThreadsafeTypeKeyHashTable<object>>();
```

Comment:
Ooh, I didn't realize in your initial proposal that you'd be caching all the resolvers. That's a memory leak for applications that create resolvers and expect them to be GC'd later, since this collection effectively prevents collection of any resolver that's ever used.
It also assumes that resolvers can't change their answers over time, which between serializations isn't a design constraint we currently have and I'd be reluctant to introduce.

I wonder though if we can still see some performance gains by caching the result of the MakeGenericMethod call itself. Would you care to try that?